### PR TITLE
fix: use ohlcv data to calculate advanced chart percentage value

### PR DESCRIPTION
**Pull request by sahar-fehri**
## **Description**

Fixes a bug where the Advanced Chart's percentage value remained static when users switched timeframes. The percentage now correctly updates to reflect the selected time range by calculating it from OHLCV data instead of legacy props.

**Problem**
When users switched timeframes in the Advanced Chart (e.g., from "Today" to "Past hour"), the descriptive text updated correctly but the percentage value remained static. This occurred because the percentage was calculated from legacy prices data that was based on a different time period than the Advanced Chart's selected time range.

**Solution**
Modified the Advanced Chart to calculate its own percentage from OHLCV data instead of using legacy props. This ensures the percentage always matches the selected time range.

`Calculation Logic`
Formula:

```
dynamicComparePrice = ohlcvData[0].close  // First candle's close price (start of time range)
dynamicPriceDiff = currentPrice - dynamicComparePrice
percentage = (dynamicPriceDiff / dynamicComparePrice) * 100
```

## **Changelog**

CHANGELOG entry: fixed percentage display on advanced charts

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/ASSETS-3031
Fixes: https://consensyssoftware.atlassian.net/browse/ASSETS-3042

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I've included tests if applicable
- [ ] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Modifies token overview header calculations and TradingView WebView visible-range initialization, which can affect price/percent display accuracy across timeframes and chart rendering. Risk is mitigated by added test coverage for key edge cases (empty data, series changes, division by zero, extra history).
> 
> **Overview**
> **Advanced token overview price header now derives its diff/% from OHLCV data instead of legacy props**, so the displayed change updates correctly when switching time ranges and matches what the chart renders (reference candle `close`). It also switches the header price/date to reflect crosshair scrubbing, and hides the diff label when no OHLCV data is available (while still guarding `0` compare prices).
> 
> **AdvancedChart now supports an explicit visible window end (`visibleToMs`)** and uses it in the WebView `timeframe`/`setVisibleRange` logic to anchor the time window to the last candle (not `Date.now()`), with additional right-side padding for the line end marker; legacy time-scale right-offset logic was removed. Tests were expanded to validate the new OHLCV-based percentage behavior and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 08cdc86e240dcacd0f96efd713bca0352a5f9629. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: app/components/UI/AssetOverview/Price/Price.advanced.tsx [hunk 1]
```diff
@@ -228,6 +224,17 @@ const PriceAdvanced = ({
 
   const dateLabel = strings(TIME_RANGE_LABELS[timeRange]);
 
+  // Calculate price diff and percentage from OHLCV data instead of legacy props
+  const dynamicComparePrice = useMemo(() => {
+    if (ohlcvData.length === 0) return null;
+    return ohlcvData[0].close; // First candle's close price = start of time range
+  }, [ohlcvData]);
```
[the remark this answers is not in the document]
Comment (sahar-fehri) [reply app/components/UI/AssetOverview/Price/Price.advanced.tsx]: [50810bc](https://github.com/MetaMask/metamask-mobile/pull/28574/commits/50810bc5575293fbf6402e9428e1dc948c1b5ae3)

**Review by Prithpal-Sooriya — DISMISSED** [review]

**Review by juanmigdr — CHANGES_REQUESTED** [review]
I can clearly see ethereum was down today and it can also be seen in the chart BUT on the percentage change I see it went up

Image

**Comment by sahar-fehri** [comment]
> Huge discrepancy that was not there before, look at the price change difference:
> 
>  Screen.Recording.2026-04-09.at.13.23.16.mov

Original code assumed ohlcvData[0] = start of selected time range which led to this descrepency, the correct approach would be to calculate the target timestamp (e.g., 24 hours ago for "1D") and search through all candles to find the one closest to that target then use that candle's price as the comparison baseline

File: app/components/UI/AssetOverview/Price/Price.advanced.tsx [hunk 2]
```diff
@@ -228,6 +225,59 @@ const PriceAdvanced = ({
...
+  const dynamicPriceDiff = useMemo(() => {
+    if (dynamicComparePrice === null) return null;
+    return currentPrice - dynamicComparePrice;
+  }, [currentPrice, dynamicComparePrice]);
+
+  // Display values: use crosshair data when hovering, otherwise use current values
+  const displayPrice = crosshairData?.close ?? currentPrice;
+
+  const displayDiff = useMemo(() => {
+    if (!dynamicComparePrice) return null;
```
[the remark this answers is not in the document]
Comment (sahar-fehri) [reply app/components/UI/AssetOverview/Price/Price.advanced.tsx]: [c767c4d](https://github.com/MetaMask/metamask-mobile/pull/28574/commits/c767c4d400f442521082b4c58b60601b70b636a8)

File: app/components/UI/AssetOverview/Price/Price.advanced.tsx [hunk 3]
```diff
@@ -247,6 +244,38 @@ const PriceAdvanced = ({
...
   const dateLabel = strings(TIME_RANGE_LABELS[timeRange]);
 
+  // Use the first candle at or after visibleFromMs — i.e. the leftmost point
+  // the line chart actually draws. The API may return extra history before the
+  // visible window; we skip those so the header matches the chart direction.
+  const dynamicComparePrice = useMemo(() => {
+    if (ohlcvData.length === 0 || visibleFromMs == null) return null;
+
+    const firstVisible =
+      ohlcvData.find((c) => c.time >= visibleFromMs) ?? ohlcvData[0];
```
Comment (sahar-fehri) [inline app/components/UI/AssetOverview/Price/Price.advanced.tsx:266]: Example for 1D: API response for 1D contains 337 bars at 15-min intervals spanning 3.5 days
but by default user is on 1D view, hence asking for only 24h data
```
visibleFromMs = lastBar.time - durationMs
              = 1775743200000 - 86400000    (86400000 = 24h in ms) (exp)
              = 1775656800000               (Apr 8 ~15:45)
```

Before the fix:
```
API data (337 bars, 3.5 days):
   Apr 7      Apr 8 AM   Apr 8 PM   Apr 9 AM   Apr 9 ~3AM Apr 9 PM
   ──────────────────── 3.5 days 

1-Widget constructor had NO timeframe option
2- TradingView calls getBars(countBack=300)
3-filterBarsForRange fallback returns last 300 bars (~3 days)
4-TradingView auto-fits to all 300 bars
5-setVisibleRange({ from: Apr8_15:45, to: Apr9_15:45 }) called BUT...
   TradingView docs: "on small screens, it will IGNORE the from parameter"
   ↓
Chart shows from midnight Apr 9 (~15h, not 24h):

   Chart viewport (what user sees):
   $1.47 ──────────────────────────── $1.49
   Apr 9 00:00                        Apr 9 15:45
   (line goes UP from left to right)

   Header percentage (from visibleFromMs Apr 8 15:45):
   dynamicComparePrice = $1.4969  (bar at Apr 8 15:45)
   currentPrice        = $1.4829
   diff = -$0.014 (-0.93%)  → shows DOWN 

   MISMATCH: chart goes UP but header says DOWN
```

After the fix:

```
Widget constructor NOW has:
   timeframe: {
     from: 1775656800  (Apr 8 15:45 in seconds),
     to:   1775743200  (Apr 9 15:45 in seconds)
   }
 
1-TradingView knows the exact window upfront
2-getBars is called with from/to matching the 24h range
3-Returns ~96 bars (Apr 8 15:45 → Apr 9 15:45)
4-TradingView auto-fits to exactly these 96 bars

   Chart viewport (what user sees):
   $1.50 ──── $1.53 ──── $1.47 ──── $1.49
   Apr 8 15:45  Apr 9 AM   Apr 9 3AM  Apr 9 15:45
   (line starts HIGH, peaks, dips, partial recovery)

   Header percentage (from visibleFromMs):
   dynamicComparePrice = $1.4969  (first bar at left edge)
   currentPrice        = $1.4829
   diff = -$0.014 (-0.93%)  → shows DOWN 

   MATCH: chart starts at $1.50, ends at $1.49 — visually DOWN
   Header says -0.93% — also DOWN
```
Adding the timeframe option in widget constructor seems to be the reliable path for mobile impl
@geositta lemme know what you think here!
Some links
https://www.tradingview.com/charting-library-docs/latest/ui_elements/Time-Scale/
https://www.tradingview.com/charting-library-docs/latest/api/interfaces/Charting_Library.ChartingLibraryWidgetOptions/#timeframe

File: app/components/UI/AssetOverview/Price/Price.advanced.tsx [hunk 4]
```diff
@@ -247,6 +244,38 @@ const PriceAdvanced = ({
...
+
+    return firstVisible.close;
+  }, [ohlcvData, visibleFromMs]);
+
+  const dynamicPriceDiff = useMemo(() => {
```
Comment (geositta) [inline app/components/UI/AssetOverview/Price/Price.advanced.tsx]: Non-blocking: dynamicPriceDiff does not seem to need to exist separately from displayDiff on line 267?

The non-crosshair path is just currentPrice minus dynamicComparePrice, which displayDiff already represents.

You could collapse this to one memo using `(crosshairData?.close ?? currentPrice) - dynamicComparePrice` and drop `dynamicPriceDiff` entirely.

Not a bug, just a simpler path for future maintainability.

**Review by geositta — DISMISSED** [review]
The OHLCV based approach here is the right direction, and the test coverage is solid. I left one comment suggestion to simplify the code a bit.

File: app/components/UI/Charts/AdvancedChart/webview/chartLogic.js [hunk 5]
```diff
@@ -3299,9 +3309,21 @@ function initChart() {
...
+    var visibleToSec = Math.ceil(
+      (window.visibleToMs != null ? window.visibleToMs : Date.now()) / 1000,
+    );
+    var tfOption =
+      window.visibleFromMs != null
+        ? {
+            from: Math.floor(window.visibleFromMs / 1000),
+            to: visibleToSec,
+          }
+        : undefined;
```
[the remark this answers is not in the document]
Comment (sahar-fehri) [reply app/components/UI/Charts/AdvancedChart/webview/chartLogic.js:3303]: [eca3c1c](https://github.com/MetaMask/metamask-mobile/pull/28574/commits/eca3c1c227b96bdf212066c985cf5a4350ebac2e)

File: app/components/UI/AssetOverview/Price/Price.advanced.tsx (hunk 4 above)
Comment (sahar-fehri) [reply app/components/UI/AssetOverview/Price/Price.advanced.tsx]: [335c98a](https://github.com/MetaMask/metamask-mobile/pull/28574/commits/335c98aa520f1e1fd96638faf7264c06f50f3c92)

File: app/components/UI/Charts/AdvancedChart/webview/chartLogic.js [hunk 6]
```diff
@@ -1416,17 +1423,20 @@ function subscribeLastCloseLabelUpdates() {
...
     ts.usePercentageRightOffset().setValue(false);
-    var gap = isLineChart
-      ? LINE_CHART_RIGHT_OFFSET_BARS
-      : CANDLE_CHART_RIGHT_GAP_BARS;
+    var gap =
+      window.visibleFromMs != null
+        ? LINE_CHART_RIGHT_OFFSET_BARS
+        : isLineChart
+          ? LINE_CHART_RIGHT_OFFSET_BARS
+          : CANDLE_CHART_RIGHT_GAP_BARS;
```
[the remark this answers is not in the document]
Comment (sahar-fehri) [reply app/components/UI/Charts/AdvancedChart/webview/chartLogic.js]: [64891ee](https://github.com/MetaMask/metamask-mobile/pull/28574/commits/64891eeebe4f2733dedab12c681fe2b9cff12fe7)

**Comment by sahar-fehri** [comment]
Refactoring chartLogic.js to help with sonar tests will be done on a separate PR

**Comment by Prithpal-Sooriya** [comment]
@juanmigdr are the image blockers you found from this PR, or blockers from the data side?

**Comment by juanmigdr** [comment]
> @juanmigdr are the image blockers you found from this PR, or blockers from the data side?

@Prithpal-Sooriya Not sure I follow, which issue exactly? Also remember these comments are all from last week, so things might have changed!

**Comment by juanmigdr** [comment]
Not a blocker (Video has audio):

https://github.com/user-attachments/assets/74d661fa-2ec0-44e8-85db-f0e100727f40

**Comment by juanmigdr** [comment]
Also not a blocker here but not sure of whats going on:

https://github.com/user-attachments/assets/8c82a2d5-8c48-4ab8-b420-058583096e38

**Comment by juanmigdr** [comment]
Proposed fix for 0% missing at the beginning of the chart:

https://github.com/user-attachments/assets/80e2e9fd-7b92-477f-99cd-4710facc1bbf

**Comment by juanmigdr** [comment]
One for the PM here:

https://github.com/user-attachments/assets/7ac49af2-ddf4-4351-96c5-709c0209cc4e

**Comment by sahar-fehri** [comment]
> Also not a blocker here but not sure of whats going on:
> 
>  Screen.Recording.2026-04-13.at.16.09.58.mov
this should be more stable here [043c6f9](https://github.com/MetaMask/metamask-mobile/pull/28574/commits/043c6f93c9aa038ec802c314f9d52e77dba3ccba)

**Comment by sahar-fehri** [comment]
> Proposed fix for 0% missing at the beginning of the chart:
> 
>  Screen.Recording.2026-04-13.at.16.13.13.mov

I tried manipulating the barspacing to try and fit in more data withing the existing timeRange; it was not working properly; will need to followup on a separate PR

**Comment by sahar-fehri** [comment]
> Not a blocker (Video has audio):
> 
>  Screen.Recording.2026-04-13.at.16.05.40.mov

This issue seems to happen whenever the head of the chart (oldest bar is visible) its also happening because of the custom CSS in place for the outline label.

File: app/components/UI/Charts/AdvancedChart/webview/chartLogic.js [hunk 7]
```diff
@@ -3418,7 +3440,6 @@ function initChart() {
                 applyChartContainerOverflowUnclip();
                 scheduleLastCloseLabelUpdate();
                 if (window.currentChartType === 2) {
-                  syncTimeScaleRightMargin(true);
```
[the remark this answers is not in the document]
Comment (sahar-fehri) [reply app/components/UI/Charts/AdvancedChart/webview/chartLogic.js:3421]: this is intentional and it does not cause clipping issues

**Review by juanmigdr — APPROVED** [review]
LGTM! Nice one
